src/index.js

```diff
@@ -0,0 +1,849 @@
+'use strict'
+
+const Pubsub = require('libp2p-pubsub')
+const FloodSub = require('libp2p-floodsub')
+
+const pull = require('pull-stream')
+const lp = require('pull-length-prefixed')
+const asyncEach = require('async/each')
+const nextTick = require('async/nextTick')
+
+const MessageCache = require('./messageCache').MessageCache
+const CacheEntry = require('./messageCache').CacheEntry
+const utils = require('./utils')
+const assert = require('assert')
+
+const RPC = require('./message').rpc.RPC
+const constants = require('./constants')
+
+
+class GossipSub extends Pubsub {
+
+    /**
+     * @param {Object} libp2p
+     * @constructor
+     */
+    constructor (libp2p) {
+        super('libp2p:gossipsub', constants.GossipSubID, libp2p)
+
+	/**
+	 * Map of topic meshes
+	 *
+	 * @type {Map<string, Set<Peer>>}
+	 */
+	this.mesh = new Map()
+
+	/**
+	 * Map of topics to lists of peers. These mesh peers are the ones to which we are publishing without a topic membership
+	 *
+	 *@type {Map<string, Set<Peer>>}
+	 */
+	this.fanout = new Map()
+
+	/**
+	 * Map of last publish time for fanout topics
+	 *
+	 *@type {Map<string, Number>}
+	 */
+	this.lastpub = new Map()
+	
+	/**
+	 * Map of pending messages to gossip
+	 *
+	 * @type {Map<Peer, Array<RPC.ControlIHave object>> }
+	 */
+	this.gossip = new Map()
+	
+	/**
+	 * Map of control messages
+	 *
+	 * @type {Map<Peer, RPC.ControlMessage object>}
+	 */
+	this.control = new Map()
+
+	/**
+	 * A message cache that contains the messages for last few hearbeat ticks
+	 *
+	 */
+	this.messageCache = new MessageCache(constants.GossipSubHistoryGossip, constants.GossipSubHistoryLength)
+	
+    }
+
+    /**
+     * Removes a peer from the router
+     * 
+     * @override
+     * @param {Peer} peer
+     * @returns {undefined}
+     */
+    _removePeer (peer) {
+        const id = peer.info.id.toB58String()
+	
+	this.log('remove', id, peer._references)
+	// Only delete when no one else if referencing this peer.
+	if (--peer._references === 0){
```

Comment:
the `--` to me seems that if `peer._references === 2` then this condition is met. It seems to me that you are actually dropping the peer if they have `2` references

Reply:
`--` is the decrement operator. If `peer._references === 2` then `--peer._references === 1`.

Reply:
Sorry yes, I meant to use `1` not `2`.

None the less, why are you performing this operation? Is it assuming that you are the only peer? i.e if `_references === 1` you are the `1` therefore `drop`?

Reply:
if `peer._references == 1` then there is 1 other peer (not equal to `peer`) referencing `peer` and so `peer` should not be dropped.

Reply:
You've lost me, assumptions:
- `this` means the current user
- `peer` is another client 

`peer._references === 1`  means that `this` is the `1` reference therefore because `this` is the only one connecting to `peer` we drop them?